Mathlib/Algebra/Lie/CartanExists.lean

```diff
@@ -0,0 +1,380 @@
+    IsRegular K x ↔ finrank K (engel K x) = rank K L := by
+  rw [isRegular_iff_natTrailingDegree_charpoly_eq_rank, finrank_engel]
+
+namespace engel_le_engel
+
+/-!
+## Implementation details for the proof of `LieAlgebra.engel_le_engel`
+
+In this section we provide some auxiliary definitions and lemmas
+that are used in the proof of `LieAlgebra.engel_le_engel`,
+which is the following statement:
+
+Let `L` be a Lie algebra of dimension `n` over a field `K` with at least `n` elements.
+Given a Lie subalgebra `U` of `L`, and an element `x ∈ U` such that `U ≤ engel K x`.
+Suppose that `engel K x` is minimal amongst the Engel subalgebras `engel K y` for `y ∈ U`.
+Then `engel K x ≤ engel K y` for all `y ∈ U`.
+
+We follow the proof strategy of Lemma 2 in [barnes1967].
+-/
+
+variable (R M)
+
+variable (x y : L)
+
+open LieModule LinearMap
+
+local notation "φ" => LieModule.toEndomorphism R L M
+
+/-- Let `x` and `y` be elements of a Lie `R`-algebra `L`, and `M` a Lie module over `M`.
+Then the characteristic polynomials of the family of endomorphisms `⁅r • y + x, _⁆` of `M`
+have coefficients that are polynomial in `r : R`.
+In other words, we obtain a polynomial over `R[X]`
+that specializes to the characteristic polynomial of `⁅r • y + x, _⁆` under the map `X ↦ r`.
+This polynomial is captured in `lieCharpoly R M x y`. -/
+private noncomputable
+def lieCharpoly : Polynomial R[X] :=
+  letI bL := chooseBasis R L
+  letI bM := chooseBasis R M
+  (LinearMap.polyCharpoly (LieHom.toLinearMap φ) bL).map <| RingHomClass.toRingHom <|
+    MvPolynomial.aeval fun i ↦ C (bL.repr y i) * X + C (bL.repr x i)
+
+lemma lieCharpoly_monic : (lieCharpoly R M x y).Monic :=
+  (polyCharpoly_monic _ _).map _
+
+lemma lieCharpoly_natDegree : (lieCharpoly R M x y).natDegree = finrank R M := by
+  rw [lieCharpoly, (polyCharpoly_monic _ _).natDegree_map, polyCharpoly_natDegree]
+
+variable {R} in
+lemma lieCharpoly_map_eval (r : R) :
+    (lieCharpoly R M x y).map (evalRingHom r) = (φ (r • y + x)).charpoly := by
+  rw [lieCharpoly, map_map]
+  set b := chooseBasis R L
+  have aux : (fun i ↦ (b.repr y) i * r + (b.repr x) i) = b.repr (r • y + x) := by
+    ext i; simp [mul_comm r]
+  simp_rw [← coe_aeval_eq_evalRingHom, ← AlgHom.comp_toRingHom, MvPolynomial.comp_aeval,
+    map_add, map_mul, aeval_C, Algebra.id.map_eq_id, RingHom.id_apply, aeval_X, aux,
+    MvPolynomial.coe_aeval_eq_eval, polyCharpoly_map_eq_charpoly, LieHom.coe_toLinearMap]
+
+lemma lieCharpoly_coeff_natDegree (i j : ℕ) (hij : i + j = finrank R M) :
+    ((lieCharpoly R M x y).coeff i).natDegree ≤ j := by
+  classical
+  rw [← mul_one j, lieCharpoly, coeff_map]
+  apply MvPolynomial.aeval_natDegree_le
+  · apply (polyCharpoly_coeff_isHomogeneous φ (chooseBasis R L) _ _ hij).totalDegree_le
+  intro k
+  apply Polynomial.natDegree_add_le_of_degree_le
+  · apply (Polynomial.natDegree_C_mul_le _ _).trans
+    simp only [natDegree_X, le_rfl]
+  · simp only [natDegree_C, zero_le]
+
+end engel_le_engel
```

Comment:
I think this entire section (from `lieCharpoly` to `lieCharpoly_coeff_natDegree`) should be done for a general family of linear maps `L →ₗ[R] Module.End R M` and moved to the file `Mathlib.Algebra.Module.LinearMap.Polynomial`.

Given a pair of endomorphisms $f$ and $g$, I think it's quite plausible that someone else will want results about the characteristic polynomial of a family $s ↦ sf + g$ at some point.

With this in mind I think it is worth the small additional effort to make the definition depend only on the two endomorphisms `f` and `g` and then add some glue to join up with the case that we have a full linear family `φ : L →ₗ[R] Module.End R M` with `f = φ x` and `g = φ y` for some `x` and `y`.

So the basic definition would be (hopefully this can be cleaned up):
```lean
-- Or `LinearMap.charpoly₂`
noncomputable def LinearMap.charpolyOfPair [Module K M] [Module.Finite K M]
    (f g : Module.End K M) : K[X][X] :=
  letI bL := chooseBasis K (Submodule.span K {f, g})
  (LinearMap.polyCharpoly (Submodule.subtype _) bL).map <| RingHomClass.toRingHom <|
    MvPolynomial.aeval fun i ↦ C (bL.repr ⟨g, Submodule.subset_span <| by simp⟩ i) * X +
      C (bL.repr ⟨f, Submodule.subset_span <| by simp⟩ i)
```

Reply:
I think that I would first want to see more evidence that this is useful in other places, if you don't mind.

Reply:
Agreed; when I wrote that I did try to think of examples and the best I could come up with were examples of situations where one cares about $\det(sf + g)$ or $\tr(sf + g)$ but I admit I couldn't think about cases (except here!) where one needed the full characteristic polynomial. I think it's totally reasonable to leave this as it stands.